crates/types/src/partitions/state.rs

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) 2023 - 2025 Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::sync::Arc;
+
+use restate_encoding::NetSerde;
+
+use crate::identifiers::PartitionId;
+use crate::logs::Lsn;
+use crate::{Merge, PlainNodeId, Version};
+
+type DashMap<K, V> = dashmap::DashMap<K, V, ahash::RandomState>;
+
+/// A map of max observed replica-set configurations for partitions
+#[derive(Clone, Default)]
+pub struct PartitionReplicaSetStates {
+    inner: Arc<Inner>,
+}
+
+#[derive(Default)]
+struct Inner {
+    partitions: DashMap<PartitionId, MembershipState>,
+}
+
+impl PartitionReplicaSetStates {
+    /// Update the membership state for a partition
+    pub fn note_observed_membership(
+        &self,
+        partition_id: PartitionId,
+        current_membership: &ReplicaSetState,
+        next_membership: &Option<ReplicaSetState>,
+    ) {
+        // we insert the partition id if unknown
+        self.inner
+            .partitions
+            .entry(partition_id)
+            .and_modify(|existing| {
+                existing.merge(current_membership, next_membership);
+            })
+            .or_insert_with(|| MembershipState {
+                observed_current_membership: current_membership.clone(),
+                observed_next_membership: next_membership.clone(),
+            });
+    }
+
+    pub fn get_membership_state(&self, partition_id: PartitionId) -> Option<MembershipState> {
+        self.inner
+            .partitions
+            .get(&partition_id)
+            .map(|k| k.value().clone())
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = (PartitionId, MembershipState)> {
```

Comment:
If we return references `(&PartitionId, &MembershipState)` then one has the option to clone them if needed.

Reply:
That would leak the underlying read lock causing users to lock the map for arbitrarily long.